### PR TITLE
Added support for PS4 dongles.

### PR DESCRIPTION
**Pull request by dynamix1337**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,6 @@
-# HID driver for Activision GH Live PS3, PS4 and Wii U Guitar devices #
+# HID driver for Activision GH Live PS3, Wii U and PS4 Guitar devices #
```
Comment (evilynux) [inline README.md]: Please use the Oxford comma here, and in the other enumerations as well.

File: README.md [hunk 2]
```diff
@@ -25,14 +25,14 @@ Without `dkms`, you will require a configured kernel source tree.
...
-3. Without `dkms`, run `cd hid-ghlive && make modules && sudo make modules_install`
+4. Without `dkms`, run `cd hid-ghlive && make modules && sudo make modules_install`
 5. Done!
 
 ### How to Update ###
 1. Move to the cloned-repository folder
 2. Update the cloned repository: `git pull`
-3. With `dkms`, run `sudo ./install.sh` or
-3. Without `dkms`, run `cd hid-ghlive && make reinstall`
+3. With `dkms`, run `sudo ./uninstall.sh` followed by `sudo ./install.sh`or
```
Comment (evilynux) [inline README.md]: Perhaps suggest `sudo ./uninstall.sh && sudo ./install.sh` instead?

File: README.md [hunk 3]
```diff
@@ -1,6 +1,6 @@
-# HID driver for Activision GH Live PS3 and Wii U Guitar devices #
+# HID driver for Activision GH Live PS3, Wii U, and PS4 Guitar devices #
 
-As far as I know this is the first module for the GH Live PS3 and Wii U Guitar devices. I really wanted to play a 6-fret guitar on [Clone Hero](https://clonehero.net/). I could not figure how to get my PS4 dongles to cooperate (yet), but thankfuly it's a different matter for the Wii U ones. Furthermore, the PS3 dongles happen to be exactly the same as the Wii U ones, so this driver supports those as well.
+This driver module supports the GH Live devices for PS3, Wii U, and PS4. This module notably allows you to play songs with the 6-fret guitar in Clone Hero with your PS3, Wii U or PS4 dongle.
```
Comment (evilynux) [inline README.md]: Missig Oxford comma ;-)

**Review by evilynux — COMMENTED** [review]
Almost there, but this won't prevent me from testing the suggested changes.

**Comment by evilynux** [comment]
Using the code from the current version of this pull request and plugging in a PS4 dongle leads to a BOGUS URB transfer:
```
[ 8996.984008] usb 1-1: BOGUS urb xfer, pipe 2 != type 1
[ 8996.984209] WARNING: CPU: 2 PID: 0 at drivers/usb/core/urb.c:493 usb_submit_urb+0x3b5/0x5e0 [usbcore]
[ 8996.984213] Modules linked in: hid_ghlive(OE) acpi_call(OE) vboxnetadp(OE) vboxnetflt(OE) vboxdrv(OE) rfcomm snd_usb_audio snd_usbmidi_lib snd_rawmidi snd_seq_device cmac algif_hash algif_skcipher af_alg usblp cdc_ether usbnet r8152 mii bnep btusb btrtl btbcm btintel bluetooth uvcvideo videobuf2_vmalloc videobuf2_memops videobuf2_v4l2 videobuf2_common jitterentropy_rng videodev sd_mod drbg sg mc nfnetlink ansi_cprng ecdh_generic ecc intel_pmc_core_pltdrv intel_pmc_core cpufreq_powersave cpufreq_userspace cpufreq_conservative cpufreq_ondemand x86_pkg_temp_thermal intel_powerclamp coretemp joydev snd_hda_codec_hdmi mei_wdt kvm_intel intel_rapl_msr iwlmvm snd_hda_codec_realtek kvm snd_hda_codec_generic snd_soc_skl irqbypass snd_soc_hdac_hda ghash_clmulni_intel mac80211 snd_hda_ext_core snd_soc_sst_ipc snd_soc_sst_dsp snd_soc_acpi_intel_match snd_soc_acpi snd_hda_intel snd_intel_dspcfg soundwire_intel soundwire_generic_allocation binfmt_misc libarc4 snd_soc_core aesni_intel libaes crypto_simd
[ 8996.984346]  snd_compress soundwire_cadence nls_ascii cryptd nls_cp437 glue_helper iwlwifi snd_hda_codec vfat iTCO_wdt rapl intel_cstate snd_hda_core intel_pmc_bxt fat iTCO_vendor_support intel_uncore serio_raw efi_pstore snd_hwdep wmi_bmof cfg80211 intel_wmi_thunderbolt rmi_smbus tpm_crb watchdog soundwire_bus rmi_core snd_pcm ee1004 snd_timer tpm_tis thinkpad_acpi tpm_tis_core nvram mei_me ledtrig_audio tpm ucsi_acpi typec_ucsi snd mei intel_xhci_usb_role_switch processor_thermal_device roles typec intel_rapl_common soundcore intel_pch_thermal intel_soc_dts_iosf rng_core rfkill int3403_thermal ac int340x_thermal_zone int3400_thermal evdev acpi_thermal_rel acpi_pad msr parport_pc ppdev lp parport fuse configfs efivarfs ip_tables x_tables autofs4 ext4 crc16 mbcache jbd2 crc32c_generic hid_generic usbhid hid uas usb_storage scsi_mod i915 i2c_algo_bit nvme drm_kms_helper e1000e xhci_pci nvme_core xhci_hcd cec t10_pi ptp crc_t10dif crct10dif_generic drm pps_core usbcore psmouse crc32_pclmul
[ 8996.984505]  crc32c_intel intel_lpss_pci intel_lpss idma64 i2c_i801 i2c_smbus crct10dif_pclmul crct10dif_common usb_common wmi battery video button [last unloaded: hid_ghlive]
[ 8996.984536] CPU: 2 PID: 0 Comm: swapper/2 Tainted: G        W  OE     5.10.0-6-amd64 #1 Debian 5.10.28-1
[ 8996.984540] Hardware name: LENOVO 20L6S4ST00/20L6S4ST00, BIOS N24ET61W (1.36 ) 10/13/2020
[ 8996.984570] RIP: 0010:usb_submit_urb+0x3b5/0x5e0 [usbcore]
[ 8996.984577] Code: 89 54 24 10 44 89 44 24 08 e8 47 02 e7 f4 44 8b 44 24 08 44 89 f9 48 8b 54 24 10 48 89 c6 48 c7 c7 38 5c 3c c0 e8 9e ce 0c f5 <0f> 0b e9 95 fd ff ff 41 83 7c 24 1c 04 0f 84 31 fd ff ff 83 e8 01
[ 8996.984581] RSP: 0018:ffffbce5c0198e78 EFLAGS: 00010286
[ 8996.984586] RAX: 0000000000000000 RBX: 0000000000000003 RCX: ffff9d640e498a08
[ 8996.984589] RDX: 00000000ffffffd8 RSI: 0000000000000027 RDI: ffff9d640e498a00
[ 8996.984592] RBP: ffff9d6265415e40 R08: 0000000000000000 R09: ffffbce5c0198c98
[ 8996.984595] R10: ffffbce5c0198c90 R11: ffffffffb62cb368 R12: ffff9d62db366000
[ 8996.984598] R13: 0000000000000003 R14: 0000000000000001 R15: 0000000000000002
[ 8996.984602] FS:  0000000000000000(0000) GS:ffff9d640e480000(0000) knlGS:0000000000000000
[ 8996.984606] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 8996.984609] CR2: 00007f6503cc0000 CR3: 00000001b9e0a003 CR4: 00000000003706e0
[ 8996.984612] DR0: 0000000000000000 DR1: 0000000000000000 DR2: 0000000000000000
[ 8996.984615] DR3: 0000000000000000 DR6: 00000000fffe0ff0 DR7: 0000000000000400
[ 8996.984618] Call Trace:
[ 8996.984624]  <IRQ>
[ 8996.984638]  ghl_magic_poke_ps4+0xec/0x140 [hid_ghlive]
[ 8996.984645]  ? ghlive_mapping+0xa0/0xa0 [hid_ghlive]
[ 8996.984656]  call_timer_fn+0x29/0xf0
[ 8996.984662]  __run_timers.part.0+0x1d3/0x240
[ 8996.984686]  ? recalibrate_cpu_khz+0x10/0x10
[ 8996.984691]  ? ktime_get+0x38/0xa0
[ 8996.984696]  ? lapic_next_deadline+0x28/0x30
[ 8996.984702]  ? clockevents_program_event+0x8d/0xf0
[ 8996.984707]  run_timer_softirq+0x26/0x50
[ 8996.984714]  __do_softirq+0xc5/0x275
[ 8996.984720]  asm_call_irq_on_stack+0x12/0x20
[ 8996.984723]  </IRQ>
[ 8996.984732]  do_softirq_own_stack+0x37/0x40
[ 8996.984740]  irq_exit_rcu+0x8e/0xc0
[ 8996.984747]  sysvec_apic_timer_interrupt+0x36/0x80
[ 8996.984755]  asm_sysvec_apic_timer_interrupt+0x12/0x20
[ 8996.984766] RIP: 0010:cpuidle_enter_state+0xc7/0x350
[ 8996.984770] Code: 8b 3d 3d 52 d7 4a e8 18 47 a2 ff 49 89 c5 0f 1f 44 00 00 31 ff e8 a9 51 a2 ff 45 84 ff 0f 85 fa 00 00 00 fb 66 0f 1f 44 00 00 <45> 85 f6 0f 88 06 01 00 00 49 63 c6 4c 2b 2c 24 48 8d 14 40 48 8d
[ 8996.984773] RSP: 0018:ffffbce5c010fea8 EFLAGS: 00000246
[ 8996.984778] RAX: ffff9d640e4abc00 RBX: 0000000000000006 RCX: 000000000000001f
[ 8996.984781] RDX: 0000000000000000 RSI: 000000003c9b199f RDI: 0000000000000000
[ 8996.984784] RBP: ffffdce5bfca2300 R08: 0000082ec6085b19 R09: 0000000000000000
[ 8996.984787] R10: 00000000000029c3 R11: 0000000000000685 R12: ffffffffb63ae5e0
[ 8996.984790] R13: 0000082ec6085b19 R14: 0000000000000006 R15: 0000000000000000
[ 8996.984802]  ? cpuidle_enter_state+0xb7/0x350
[ 8996.984809]  cpuidle_enter+0x29/0x40
[ 8996.984816]  do_idle+0x1ef/0x2b0
[ 8996.984823]  cpu_startup_entry+0x19/0x20
[ 8996.984830]  secondary_startup_64_no_verify+0xb0/0xbb
[ 8996.984837] ---[ end trace 0b403e9d36652e3f ]---
```

File: hid-ghlive/src/hid-ghlive.c [hunk 4]
```diff
@@ -100,6 +115,46 @@ static void ghl_magic_poke(struct timer_list *t)
...
 
+static void ghl_magic_poke_ps4(struct timer_list *t)
+{
+	int ret;
+	struct urb *urb;
+	struct ghlive_sc *sc = from_timer(sc, t, poke_timer);
+	struct usb_device *usbdev = to_usb_device(sc->hdev->dev.parent->parent);
+	const u16 poke_size =
+		ARRAY_SIZE(ghl_ps4_magic_data);
+	unsigned int pipe = usb_sndctrlpipe(usbdev, 0x02);
```
Comment (evilynux) [inline hid-ghlive/src/hid-ghlive.c]: Should that be `unsigned int pipe = rcvctrlpipe(usbdev, 0x02);` ?

Comment (evilynux) [reply hid-ghlive/src/hid-ghlive.c]: Nevermind, that wouldn't make sense, it has to go OUT.

Comment (evilynux) [reply hid-ghlive/src/hid-ghlive.c]: If that helps, here's my device descriptor:
```
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1430 RedOctane
  idProduct          0x07bb 
  bcdDevice            1.10
  iManufacturer           1 Activision
  iProduct                2 Guitar Hero 
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0029
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xc0
      Self Powered
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     160
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               5
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x02  EP 2 OUT
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               5
```

File: README.md (hunk 3 above)
Comment (dynamix1337) [reply README.md]: Oxford comma is used for OR as well as AND?

File: hid-ghlive/src/hid-ghlive.c (hunk 4 above)
Comment (dynamix1337) [reply hid-ghlive/src/hid-ghlive.c]: What are you running in order to see the bogus?

File: README.md (hunk 2 above)
Comment (dynamix1337) [reply README.md]: Agreed. Done.

File: README.md (hunk 3 above)
Comment (dynamix1337) [reply README.md]: Just read up on Oxford commas. All good. Makes sense!

File: hid-ghlive/src/hid-ghlive.c (hunk 4 above)
Comment (dynamix1337) [reply hid-ghlive/src/hid-ghlive.c]: Actually, i forgot to change the function from a Control to an Interrupt.

`unsigned int pipe = usb_sndctrlpipe(usbdev, 0x02);` should be `unsigned int pipe = usb_sndintpipe(usbdev, 0x02);`

Comment (evilynux) [reply hid-ghlive/src/hid-ghlive.c]: Indeed. That fixes it. Please include this change when you'll push the other things discussed on Slack.

**Pull request merged**